---
title: "lymphoma cohort - tail genes"
author: "Annelien Morlion"
date: '2023-05-18'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(tidyverse)
require(ggrepel)
library(here)
library(DESeq2)

## Define plot style for paper
mytheme = theme_classic(base_size = 7) +
  theme(text = element_text(size=7, colour="black"),
        title = element_text(size=7, colour="black"),
        line = element_line(size=0.5),
        axis.title = element_text(size=7, colour="black"),
        axis.text = element_text(size=7, colour="black"),
        axis.ticks = element_line(size=0.5),
        strip.background = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        legend.position = c(0.8,0.8),
        legend.text = element_text(size=6))
mytheme_discrete_x = mytheme + theme(axis.text.x = element_text(angle=90, hjust=1,vjust=0.5))

color_panel<-c("#e35d6a","#428bca","#5bb75b","#e87810","#23496b","#ffbf00","#cc2028","#039748","pink","gray","darkgray")
cb_color_panel <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

#see https://personal.sron.nl/~pault/
cb_color_panel_highcontrast <- c("#000000","#004488", "#DDAA33", "#BB5566") #+"#FFFFFF" white
cb_color_panel_mediumcontrastpairs <- c('#6699CC', '#004488', '#EECC66', '#994455', '#997700', '#EE99AA') #+"#FFFFFF" white and "#000000" black
cb_color_panel_bright <- c('#4477AA', '#EE6677', '#228833', '#CCBB44', '#66CCEE', '#AA3377', '#BBBBBB')

full_nr <- scales::format_format(big.mark = ",", decimal.mark = ".", scientific = FALSE)

##sample annotation
sample_annotation <- data.table::fread("../input/SupplTable1_sample_annotation.txt", sep="\t", quote="",header=T, data.table=FALSE) %>% mutate(UniqueID = paste0(Abbreviation,"_",ReplicateNr,"_",Cohort))
sample_annotation_all <- sample_annotation # keep original table 
sample_annotation <- sample_annotation_all %>% filter(Excluded != "x") #filter out low read samples
```

Normalize data (only needs to be run first time)
```{r, eval=F}
lymphoma_counts <- data.table::fread("../input/lymphoma_counts.txt", data.table = F, header=T) %>%
  filter(gene_id %in% grep("__",gene_id, invert=T, value=T)) %>%
  pivot_longer(names_to="RNAID",values_to="counts",-"gene_id") %>%
  right_join(sample_annotation %>% filter(Cohort=="lymphoma") %>% dplyr::select("RNAID", "UniqueID"), by="RNAID") #leave out samples with almost no reads

library(biomaRt)
library(DESeq2)
ensembl <- useEnsembl( biomart="ensembl",dataset="hsapiens_gene_ensembl",version=91)
genes_ens <- getBM(attributes=c('ensembl_gene_id','gene_biotype'),mart=ensembl) #get gene biotype

pc_gene_counts <- lymphoma_counts %>% left_join(., genes_ens, by=c("gene_id"="ensembl_gene_id")) %>% #add gene biotype
  filter(gene_biotype=="protein_coding") %>% #only keep protein coding genes
  dplyr::select(-c("gene_biotype","RNAID")) %>% #remove gene biotype column, RNAID
  pivot_wider(names_from=UniqueID, values_from=counts) #back to original format: rows=genes, columns=samples

normal_samples <- sample_annotation %>% filter(Cohort=="lymphoma") %>% filter(Abbreviation=="CONTROL") %>% pull(UniqueID)
normal_counttable <- dplyr::select(pc_gene_counts, c("gene_id",(normal_samples)))

tumor_samples <- sample_annotation %>% filter(Cohort=="lymphoma") %>% filter(Abbreviation!="CONTROL") %>% pull(UniqueID)
tumor_counttable <- dplyr::select(pc_gene_counts, c("gene_id",(tumor_samples))) 
counttable <- full_join(normal_counttable, tumor_counttable, by="gene_id") #join tumor and normal samples
rm(tumor_counttable, normal_counttable)
  
  ## FILTER genes
keep <- (rowSums(dplyr::select(counttable, all_of(normal_samples)) > 9) >= round(length(normal_samples)/2,0)) | #≥50% of normal samples count > 9
  (rowSums(dplyr::select(counttable, all_of(tumor_samples)) > 9) >= round(length(tumor_samples)/2,0)) # OR ≥50% of tumor samples count > 9
#table(keep, useNA="ifany")
#counttable <- counttable[keep, ]
counttable <- column_to_rownames(counttable, "gene_id")

condition <- factor(ifelse(grepl("^CONTROL", colnames(counttable)), "Normal", "Tumor"))
coldata <- data.frame(row.names=colnames(counttable), condition)
dds <- DESeqDataSetFromMatrix(countData=counttable, colData=coldata, design=~condition)

dds <- DESeq(dds)
  
counttable_normalized<-as.data.frame(counts(dds, normalized = TRUE)) %>% rownames_to_column("gene_id")
data.table::fwrite(counttable_normalized, file="../data/lymphoma_normcounts.txt", quote=F, na="", row.names = F, sep="\t")
rm(counttable_normalized, dds, coldata,condition, counttable,keep,pc_gene_counts, ensembl, genes_ens, normal_samples, tumor_samples)

```

Tail gene analysis: z-score calculation function
```{r}
### log transformation + z scores calculation based on mean and stdev of (log transformed) reference distribution
tail_z_score <- function(samples, Ref_matrix, na.rm=T) {
  #samples: (normalized, non-log transformed) count matrix of samples of interest (genes as rownames, samples as columns)
  #N_matrix: (normalized, non-log transformed) count matrix of reference (genes as rownames, samples as columns) 
  require(tidyverse)
  #logtransform reference
  Ref_logmatrix <- log2(Ref_matrix+1)
  Ref_zscores <- t(Ref_logmatrix) %>% scale(center=T,scale=T) #column scaling and centering (per gene after transformation)
  ##check: mean per gene = 0, sd = 1
  #Ref_zscores %>% data.frame() %>% rownames_to_column("sample") %>% pivot_longer(names_to="gene", values_to = "counts", -"sample") %>% group_by(gene) %>% dplyr::summarise(mean_x=mean(counts), sd_x=sd(counts)) %>% View()
  scaling_factors <- data.frame(mean_Ref = attr(Ref_zscores, "scaled:center"), sd_Ref = attr(Ref_zscores, "scaled:scale")) %>% rownames_to_column("gene_id")
  #mns <- scaling_factors %>% pull(mean_Ref) #mean
  #sds <- scaling_factors %>% pull(sd_Ref) #stdev
  
  #logtransform samples and subtract mean and stdev of reference matrix
  x <- log2(samples+1) %>% data.frame()
  #match returns a vector of the positions of (first) matches of its first argument in its second -> retrieves the corresponding scaling factors for these genes
  mns <- scaling_factors$mean_Ref[match(rownames(x), scaling_factors$gene_id)] 
  sds <- scaling_factors$sd_Ref[match(rownames(x), scaling_factors$gene_id)]
  
  x <- sweep(x,1,mns,"-", check.margin = T) #subtract corresponding reference gene mean from gene counts
  x <- sweep(x,1,sds,"/", check.margin = T) #then divide gene counts by corresponding reference gene stdev
  x
}

```

calculate z-scores for each sample, each time using all control samples as reference except the control sample of interest
```{r}
lymphoma_normcounts <- data.table::fread("../data/lymphoma_normcounts.txt")

# z-score calculation based on leave one out (LOO) reference in case sample of interest is control sample
z_scores_LOO <- data.frame(gene_id = lymphoma_normcounts %>% pull(gene_id))
for (samp in (colnames(lymphoma_normcounts)[-1])) {
  #print(samp)
  z_scores_LOO_sample <- tail_z_score(lymphoma_normcounts %>% dplyr::select(c("gene_id",all_of(samp))) %>% column_to_rownames("gene_id"),
                                                                lymphoma_normcounts %>% dplyr::select(c("gene_id",grep("CONTROL",colnames(lymphoma_normcounts),value =T))) %>% 
                                                  dplyr::select(-any_of(samp)) %>% #remove sample of interest from reference
                                                  column_to_rownames("gene_id"))
  
  z_scores_LOO <- left_join(z_scores_LOO, z_scores_LOO_sample %>% rownames_to_column("gene_id"), by="gene_id")
}

data.table::fwrite(z_scores_LOO, file = "../data/lymphoma_zscores_tail.txt", sep="\t",na="",quote=F, row.names = F)

### save table with tail genes per sample (|z|>3 and ncounts ≥40)
tail_genes <- left_join(z_scores_LOO %>% pivot_longer(names_to="sample", values_to="zscores",-"gene_id") %>% filter(!is.na(zscores)) %>% filter(abs(zscores)>3), 
                  lymphoma_normcounts %>% #add normalized counts
                   pivot_longer(names_to="sample",values_to="ncounts",-"gene_id"),
                  by=c("gene_id"="gene_id","sample")) %>% filter(ncounts>=40) 
#summary(tmp)
data.table::fwrite(tail_genes, file="../data/lymphoma_tailgenes.txt",sep="\t", row.names = F, quote=F, na="")
rm(tail_genes,z_scores_LOO_sample, samp, z_scores_LOO)

```

Number of tail genes (ncounts≥40, |z|>3) and heatmap
```{r}
z_scores_LOO <- data.table::fread("../data/lymphoma_zscores_tail.txt",data.table = F)
tail_genes <- data.table::fread("../data/lymphoma_tailgenes.txt",data.table=F)
length(unique(tail_genes$gene_id)) 

require(pheatmap)
require(RColorBrewer)

#show all genes that are tail gene in at least one sample
tmp2 <- z_scores_LOO %>% filter(gene_id %in% (tail_genes %>% pull(gene_id) %>% unique()))
table(is.na(tmp2))
tmp2[is.na(tmp2)] <- 0 #0/0 = NA

#top off z-scores (max -5 to 5)
tmp2 <- tmp2 %>% column_to_rownames("gene_id")
tmp2[tmp2>5] <- 5
tmp2[tmp2<(-5)] <- -5
tmp2 <- tmp2 %>% rownames_to_column("gene_id")
#table(tmp2==Inf)

paletteLength=40
myColor <- colorRampPalette(c("blue", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd palette lengths
myBreaks <- c(seq(min(tmp2%>% column_to_rownames("gene_id")), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(tmp2%>% column_to_rownames("gene_id"))/paletteLength, max(tmp2%>% column_to_rownames("gene_id")), length.out=floor(paletteLength/2)))

myBreaks <- c(seq(min(tmp2%>% column_to_rownames("gene_id")), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(tmp2%>% column_to_rownames("gene_id"))/paletteLength, max(tmp2%>% column_to_rownames("gene_id")), length.out=floor(paletteLength/2)))

# no clustering of columns, only rows
colnames(tmp2) <- sub("CONTROL","CONTROL",sub("_([0-9])$","_0\\1",gsub("_lymphoma","",colnames(tmp2)))) #add trailing 0 for values below 10
tmp2 <- dplyr::select(tmp2, c("gene_id", colnames(tmp2)[-1] %>% sort()))
annot_cols= data.frame(type = gsub("_.*","",colnames(tmp2)), 
        row.names = colnames(tmp2))
annot_cols$type <- factor(annot_cols$type, levels=c("CONTROL","DLBCL","PMBCL"))

#differential abundance annotation
DAA_genes <- data.table::fread("../data/differentialabundance_lymphoma.txt", data.table = F) %>% mutate(DA=ifelse((padj<0.05) & (log2FoldChange>1), "higher", ifelse((padj<0.05) & (log2FoldChange<(-1)), "lower", "no")))  #higher/lower: differentially abundant genes (DA)
annot_rows = data.frame(namesrow=tmp2$gene_id) %>% 
  mutate(differential=ifelse(namesrow %in% (DAA_genes %>% filter(DA=="lower") %>% pull(gene_id)),"lower",
                             ifelse(namesrow %in% (DAA_genes %>% filter(DA=="higher") %>% pull(gene_id)),"higher","no"))) %>% column_to_rownames("namesrow")

annoCol<-list(type=c(CONTROL="#DDAA33",DLBCL="#BB5566",PMBCL="#004488"), differential=c(lower="#CCBB44",higher="#AA3377",no="#FFFFFF"))

pheatmap(as.matrix(tmp2 %>% column_to_rownames("gene_id")), 
         color=myColor, breaks=myBreaks, annotation_col = annot_cols,
         annotation_row=annot_rows,border_color=NA,cluster_cols = F,
         annotation_colors = annoCol, annotation_names_col=F, annotation_names_row = F,
         cluster_rows=T,fontsize = 6,fontsize_row = 7, fontsize_col=7,
         show_colnames = F, scale="none", show_rownames = F, clustering_distance_rows = "euclidean",
         treeheight_row = 0, gaps_col = head(as.numeric(cumsum(table(annot_cols$type))),-1), angle_col = 90,
         legend_breaks = c(-4, -2, 0, 2, 4, max(tmp2 %>% column_to_rownames("gene_id"))), 
         legend_labels = c("-4","-2","0","2","4", "z  \n"),#add legend title
         filename = "../figures/lymphoma_tail_heatmap.pdf", 
         width=4.76, height=2.40) #inches (width 121mm=4.76in, height 61mm=2.40in)

png(filename="../figures/lymphoma_tail_heatmap.png",res = 300, width=12.1, height=6.1, units = "cm")
pheatmap(as.matrix(tmp2 %>% column_to_rownames("gene_id")), 
         color=myColor, breaks=myBreaks, annotation_col = annot_cols,
         annotation_row=annot_rows,border_color=NA,cluster_cols = F,
         annotation_colors = annoCol, annotation_names_col=F, annotation_names_row = F,
         cluster_rows=T,fontsize = 6,fontsize_row = 7, fontsize_col=7,
         show_colnames = F, scale="none", show_rownames = F, clustering_distance_rows = "euclidean",
         treeheight_row = 0, gaps_col = head(as.numeric(cumsum(table(annot_cols$type))),-1), angle_col = 90,
         legend_breaks = c(-4, -2, 0, 2, 4, max(tmp2 %>% column_to_rownames("gene_id"))), 
         legend_labels = c("-4","-2","0","2","4", "z  \n"))#add legend title
dev.off()

rm(tmp2,annoCol,annot_rows,annot_cols,DAA_genes, myBreaks,paletteLength)

```


Recurrence of tail genes
```{r}
nr_samples_total <- sample_annotation %>% filter(Cohort=="lymphoma") %>% group_by(Abbreviation) %>% dplyr::summarise(nr_samples_total=n()) 
tmp2 <- tail_genes %>% mutate(Abbreviation=gsub("_.*","",sample)) %>% group_by(gene_id,Abbreviation) %>% dplyr::summarise(nr_dev_samples=n()) %>% ungroup() %>% left_join(nr_samples_total, by="Abbreviation") %>% mutate(perc_dev_samples=nr_dev_samples/nr_samples_total)
#make sure all rows are represented (0 if absent)
tmp3 <- tmp2 %>% ungroup() %>% group_by(Abbreviation,nr_dev_samples) %>% dplyr::summarise(nr_genes=n()) %>% full_join(data.frame(Abbreviation=c(rep("CONTROL",22), rep("DLBCL",30), rep("PMBCL",13)),nr_dev_samples=c(seq(1:22), seq(1:30), seq(1:13) )))

tmp3[is.na(tmp3)] <- 0

ggplot(tmp3, aes(x=nr_dev_samples, y=nr_genes+0.05,fill=Abbreviation)) +
  geom_bar(stat="identity",position="dodge", width=0.8) +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488")) +
  scale_y_log10(expand=c(0,NA),limits=c(0.9,NA)) +
  mytheme + theme(legend.title = element_blank()) +
  labs(x="number of deviating samples", y="number of tail genes")
ggsave("../figures/lymphoma_tail_recurrence.pdf", plot=last_plot(), height=6.1, width=5.8, units="cm",  dpi = 300, useDingbats=F)

tmp3 <- tmp2 %>% ungroup() %>% group_by(Abbreviation,perc_dev_samples) %>% dplyr::summarise(nr_genes=n())
tmp3 <- tmp3 %>% 
  full_join(tmp3 %>% group_by(Abbreviation) %>% dplyr::summarise(perc_dev_samples=max(perc_dev_samples)+0.05)) %>% #add a row with the max perc + 5%
  full_join(data.frame(cancer=c("CONTROL","DLBCL","PMBCL"), perc_dev_samples=c(1,1,1))) #add rows for 100%
tmp3$nr_genes[is.na(tmp3$nr_genes)] <- 0 #fill in 0 deviating genes for missing values

ggplot(tmp3, aes(x=perc_dev_samples*100, y=nr_genes,color=Abbreviation)) +
  geom_line() + geom_point(size=0.5) + 
  scale_color_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488")) +
  scale_y_log10() +
  scale_x_continuous() +
  mytheme + theme(legend.title = element_blank()) +
  labs(x="shared among % of samples", y="number of tail genes")
ggsave("../figures/lymphoma_tail_recurrence_perc.pdf", plot=last_plot(), height=6.1, width=5.8, units="cm",  dpi = 300, useDingbats=F)


tmp4 <- tmp2 %>% ungroup() %>% mutate(CN=ifelse(Abbreviation=="CONTROL","control","cancer")) %>% 
  group_by(gene_id,CN) %>% dplyr::summarise(nr_dev_samples_allcancer=sum(nr_dev_samples)) %>% #sum nr dev cancer samples
  ungroup() %>% group_by(CN,nr_dev_samples_allcancer) %>% dplyr::summarise(nr_genes=n())
#nr of tail genes not unique for 1 sample
(tmp4 %>% filter(CN=="cancer") %>% filter(nr_dev_samples_allcancer!=1) %>% pull(nr_genes) %>% sum())/(tmp4 %>% filter(CN=="cancer") %>% pull(nr_genes) %>% sum())
(tmp4 %>% filter(CN=="control") %>% filter(nr_dev_samples_allcancer!=1) %>% pull(nr_genes) %>% sum())/(tmp4 %>% filter(CN=="control") %>% pull(nr_genes) %>% sum())

tmp4 <- tmp2 %>% 
  group_by(gene_id,Abbreviation) %>% dplyr::summarise(nr_dev_samples_allcancer=sum(nr_dev_samples)) %>% #sum nr dev cancer samples
  ungroup() %>% group_by(Abbreviation,nr_dev_samples_allcancer) %>% dplyr::summarise(nr_genes=n())
(tmp4 %>% filter(Abbreviation=="DLBCL") %>% filter(nr_dev_samples_allcancer!=1) %>% pull(nr_genes) %>% sum())/(tmp4 %>% filter(Abbreviation=="DLBCL") %>% pull(nr_genes) %>% sum())
(tmp4 %>% filter(Abbreviation=="PMBCL") %>% filter(nr_dev_samples_allcancer!=1) %>% pull(nr_genes) %>% sum())/(tmp4 %>% filter(Abbreviation=="PMBCL") %>% pull(nr_genes) %>% sum())

### overlap between subtypes
tmp3 <- tail_genes %>% mutate(Abbreviation=gsub("_.*","",sample)) %>% group_by(gene_id,Abbreviation) %>% dplyr::summarise(nr_dev_samples=n()) %>% ungroup() %>% left_join(nr_samples_total, by="Abbreviation") %>% mutate(perc_dev_samples=nr_dev_samples/nr_samples_total)

rm(tmp2,tmp3,tmp4,nr_samples_total,tail_genes)
```

Overlap of tail gene sets
```{r}
tail_genes <- data.table::fread("../data/lymphoma_tailgenes.txt",data.table=F) %>%
   mutate(Abbreviation=gsub("_.*","",sample)) #deviating genes (Tail)

### overlap plot of tail genes
library(eulerr)
fit <- eulerr::euler(list("DLBCL"=tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique(), "PMBCL"=tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique()), 
                       shape="ellipse")
p1<- plot(fit, quantities = TRUE,shape="ellipse", alpha=0.7,edges="white", legend=F,fills=c(DLBCL="#BB5566",PMBCL="#004488"), main="overlap tail genes")
print(fit) #no residual errors 
print(p1)

ggsave("../figures/lymphoma_tailoverlap.pdf", plot=p1, useDingbats=F,
       width=5.8, height=6.1, units="cm")

#include control group tail genes
fit <- eulerr::euler(list("DLBCL"=tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique(), "PMBCL"=tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique(), "CONTROL"=tail_genes %>% filter(Abbreviation=="CONTROL") %>% pull(gene_id) %>% unique()), 
                       shape="ellipse")
p1<- plot(fit, quantities = TRUE,shape="ellipse", alpha=0.7, #make it a bit more light
          legend=F,fill=c(DLBCL="#BB5566",PMBCL="#004488",CONTROL="#DDAA33"), main="overlap tail genes")
print(p1)
print(fit) #residual errors with 4 groups!

#significance of overlap
jaccard <- function(a, b) { #calculate jaccard index
    intersection = length(intersect(a, b))
    union = length(a) + length(b) - intersection
    return (intersection/union)
}

require(GeneOverlap)
#universe?
#how many unique genes have at least 40 normalized counts in 1 sample
universe_lymphoma <- length(lymphoma_normcounts %>% pivot_longer(names_to="sample",values_to="ncounts",-"gene_id") %>% filter(ncounts>=40) %>% pull(gene_id) %>% unique()) #7649 genes with at least 1 sample with ≥40 normalized counts

require(GeneOverlap)
#test overlap via GeneOverlap object
go.obj.DLPM <- newGeneOverlap(tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique(),
                         genome.size=universe_lymphoma) #all (5373) tail genes?
go.obj.DLPM <- testGeneOverlap(go.obj.DLPM)
#The GeneOverlap class formulates the problem as testing whether two variables are independent, which can be represented as a contingency table, and then uses Fisher’s exact test to find the statistical significance. Here the P-value is zero, which means the overlap is highly significant. The Fisher’s exact test also gives an odds ratio which represents the strength of association. If an odds ratio is equal to or less than 1, there is no association between the two lists. If the odds ratio is much larger than 1, then the association is strong. The class also calculates the Jaccard index which measures the similarity between two lists. The Jaccard index varies between 0 and 1, with 0 meaning there is no similarity between the two and 1 meaning the two are identical. To show some more details, use the print function
print(go.obj.DLPM)
jaccard(tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique())

#overlap with control tail genes?
go.obj.CTRLDL <- newGeneOverlap(tail_genes %>% filter(Abbreviation=="CONTROL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique(),
                         genome.size=universe_lymphoma) 
go.obj.CTRLDL <- testGeneOverlap(go.obj.CTRLDL)
print(go.obj.CTRLDL)
jaccard(tail_genes %>% filter(Abbreviation=="CONTROL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="DLBCL") %>% pull(gene_id) %>% unique())

go.obj.CTRLPM <- newGeneOverlap(tail_genes %>% filter(Abbreviation=="CONTROL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique(),
                         genome.size=universe_lymphoma) 
go.obj.CTRLPM <- testGeneOverlap(go.obj.CTRLPM)
print(go.obj.CTRLPM)
jaccard(tail_genes %>% filter(Abbreviation=="CONTROL") %>% pull(gene_id) %>% unique(), tail_genes %>% filter(Abbreviation=="PMBCL") %>% pull(gene_id) %>% unique())

```

test whether there are significantly less tail genes in certain groups
```{r}
tail_genes <- data.table::fread("../data/lymphoma_tailgenes.txt",data.table=F) %>%
   mutate(disease=gsub("_.*","",sample)) #deviating genes (Tail) 

tmp2 <- tail_genes %>% group_by(sample) %>% dplyr::summarise(n_dev = n()) %>% mutate(disease=gsub("_.*","",sample))
#make sure no sample is missing
tmp2 <- tmp2 %>% full_join(sample_annotation %>% filter(Cohort=="lymphoma") %>% dplyr::select(c("sample"=UniqueID,"disease"=Abbreviation)), by=c("sample","disease")) %>% mutate(n_dev=ifelse(is.na(n_dev),0,n_dev))

library(ggpubr)
#my_comparisons <- list( c("CONTROL", "DLBCL"), c("CONTROL", "PMBCL") )
ggplot(tmp2, aes(x=disease,y=n_dev)) +
  geom_boxplot(outlier.size = 0.75,aes(fill=disease)) +
  geom_jitter(height=0, width=0.1,size=0.5,color="grey")+
  mytheme + theme(legend.position="none") +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488"))+
  scale_y_log10(labels=full_nr) +
  labs(y="tail genes",x="") +
  ggpubr::stat_compare_means(method="kruskal",label.y=4, size = 2)   # Add global p-value
  #ggpubr::stat_compare_means(comparisons=my_comparisons, method = "wilcox.test", size = 2) # Pairwise comparison against reference
#ggpubr::compare_means(n_dev ~ disease, data = tmp2,method = "wilcox.test",p.adjust.method = "BH")

rstatix::kruskal_test(data=tmp2, n_dev ~ disease)
rstatix::kruskal_effsize(data=tmp2, n_dev ~ disease)
#rstatix::kruskal_effsize(data=tmp2, n_dev ~ disease,ci=T)
rstatix::wilcox_test(data=tmp2, n_dev ~ disease, p.adjust.method = "BH")
rstatix::wilcox_effsize(data=tmp2, n_dev ~ disease)
#rstatix::wilcox_effsize(data=tmp2, n_dev ~ disease, ci = T)

ggsave("../figures/lymphoma_tail_boxplot_logscale.pdf", plot=last_plot(), height=6.1, width=5.8, units="cm",  dpi = 300, useDingbats=F)
kruskal.test(n_dev ~ disease, data = tmp2) # significant difference between Abbreviation types
#pairwise.wilcox.test(tmp2$n_dev, tmp2$disease,p.adjust.method = "BH")
#pairwise.wilcox.test(tmp2$n_dev, tmp2$disease,p.adjust.method = "holm")

ggplot(tmp2, aes(x=disease,y=n_dev)) +
  geom_boxplot(outlier.size = 0.75,aes(fill=disease)) +
  geom_jitter(height=0, width=0.1,size=0.5,color="grey")+
  mytheme + theme(legend.position="none") +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488"))+
  #scale_y_log10(labels=full_nr) +
  labs(y="tail genes",x="") +
  ggpubr::stat_compare_means(method="kruskal",label.y=1400, size = 2)   # Add global p-value
  #ggpubr::stat_compare_means(comparisons=my_comparisons, method = "wilcox.test", size = 2) # Pairwise comparison against reference

rstatix::kruskal_test(data=tmp2, n_dev ~ disease)
rstatix::kruskal_effsize(data=tmp2, n_dev ~ disease)
#rstatix::kruskal_effsize(data=tmp2, n_dev ~ disease,ci=T)
rstatix::wilcox_test(data=tmp2, n_dev ~ disease, p.adjust.method = "BH")
rstatix::wilcox_effsize(data=tmp2, n_dev ~ disease)
#rstatix::wilcox_effsize(data=tmp2, n_dev ~ disease, ci = T)

ggsave("../figures/lymphoma_tail_boxplot.pdf", plot=last_plot(), height=6.1, width=5.8, units="cm",  dpi = 300, useDingbats=F)
kruskal.test(n_dev ~ disease, data = tmp2) # significant difference between Abbreviation types

```


Fisher exact test (all samples except one) to make selection of tail genes specifically associated to cancer/control state, then build model on consensus genes only (also at this level LOO with cross-validation: test model on remaining sample)
```{r}
predictions <- c()
original_type <- c()

z_scores_LOO <- data.table::fread("../data/lymphoma_zscores_tail.txt",data.table=F) 
colnames(z_scores_LOO) <- gsub("_lymphoma","",colnames(z_scores_LOO))
tail_genes <- data.table::fread("../data/lymphoma_tailgenes.txt",data.table=F) %>%
   mutate(disease=gsub("_.*","",sample)) #Tail genes

#subset z scores table to genes that are Tail genes in at least 1 sample
z_scores_LOO_Tail <- z_scores_LOO %>% filter(gene_id %in% unique(tail_genes$gene_id)) 

consensus_genes_100 <- list()
gene_list <- vector(mode="list", length=length(colnames(z_scores_LOO_Tail)[-1]))
### Find biomarker tail genes for (all) cancer vs control
for (samplename in colnames(z_scores_LOO_Tail)[-1]){
  print(samplename)
  reduced_z_table <- z_scores_LOO_Tail %>% dplyr::select(-paste(samplename))
  
  #Fisher exact test on all samples except sample of interest
  test <- reduced_z_table %>% pivot_longer(names_to="sample", values_to="zscores",-"gene_id") %>% #filter(!is.na(zscores)) %>% 
                   #filter(abs(zscores)>3) %>%
  mutate(disease=gsub("_.*","",sample)) %>% mutate(cancer=ifelse(disease=="CONTROL", "N", "T"))  %>% #get tumor/normal annotation
  #mutate(abszabove3=ifelse(abs(zscores)>3, "above3","below3"))
  ungroup() %>% group_by(gene_id,cancer) %>% dplyr::summarise(absz3 = sum(abs(zscores)>3, na.rm = T), absznot3 = sum(abs(zscores)<=3, na.rm=T))
  test_wide <- test %>% distinct() %>% pivot_wider(names_from=cancer, values_from=c(absz3,absznot3)) %>% column_to_rownames("gene_id")
  df_fisher <- data.frame(p.val=apply(test_wide,1, function(x) fisher.test(matrix(as.numeric(x[1:4]), ncol=2, byrow=T))$p.value))
#df_fisher$p.adj = p.adjust(df_fisher$p.val, method = "BH")
  df_fisher %>% filter(p.val<0.05) %>% nrow()
  gene_list[[samplename]] <- df_fisher %>% filter(p.val<0.05) %>% rownames()
}

#count how many times a gene is significant in fisher test of all samples except 1 (65 repeats in total so max is 65)
tmp <-data.frame("gene_id"=c(unlist(gene_list))) %>% group_by(gene_id) %>% 
  dplyr::summarise(nr_occ=n())
#ggplot(tmp, aes(x=nr_occ)) +geom_bar() + theme_point

#Keep only genes that are significant in ALL sets to build model (+LOO crossval)
nr_sets = length(colnames(z_scores_LOO_Tail)[-1])
consensus_genes_100[["all"]] <- tmp %>% filter(nr_occ==nr_sets) %>% pull(gene_id) #1113 consensus genes

### Find biomarker tail genes for every (individual) cancer-control comparison
for (cancertype in c("DLBCL","PMBCL")){
  #make a reduced z_table with only z_score gene
  Tail_GOI <- tail_genes %>% filter(disease %in% c("CONTROL",cancertype)) %>% pull(gene_id) %>% unique()#tail genes of interest (Tail gene in at least one of the cancer or control samples)
  #make a reduced z score table
  cancer_z_table <- z_scores_LOO %>% filter(gene_id %in% Tail_GOI) %>% #filter for Tail GOI 
    dplyr::select(c("gene_id",grep(paste0(cancertype,"|CONTROL"),colnames(z_scores_LOO),value = T)))
  
  
  #Fisher's Exact test on all samples except one within cancer-control comparison
  gene_list <- vector(mode="list", length=length(colnames(cancer_z_table)[-1]))
  names(gene_list) <- colnames(cancer_z_table)[-1]
  for (samplename in colnames(cancer_z_table)[-1]){
    print(samplename)
    reduced_z_table <- cancer_z_table %>% dplyr::select(-paste(samplename))
    #Fisher exact test on all samples except sample of interest
    test <- reduced_z_table %>% 
      pivot_longer(names_to="sample", values_to="zscores",-"gene_id") %>%
      mutate(disease=gsub("_.*","",sample)) %>% 
      mutate(cancer=ifelse(disease=="CONTROL", "N", "T"))  %>% #get tumor/normal annotation
      ungroup() %>% group_by(gene_id,cancer) %>% 
      dplyr::summarise(absz3 = sum(abs(zscores)>3, na.rm = T), absznot3 = sum(abs(zscores)<=3, na.rm=T))
  
    test_wide <- test %>% distinct() %>% pivot_wider(names_from=cancer, values_from=c(absz3,absznot3)) %>% column_to_rownames("gene_id")
    df_fisher <- data.frame(p.val=apply(test_wide,1, function(x) fisher.test(matrix(as.numeric(x[1:4]), ncol=2, byrow=T))$p.value))
#df_fisher$p.adj = p.adjust(df_fisher$p.val, method = "BH")
    df_fisher %>% filter(p.val<0.05) %>% nrow()
    gene_list[[samplename]] <- df_fisher %>% filter(p.val<0.05) %>% rownames()
  }
  head(unlist(gene_list))
  #count how many times a gene is significant in fisher test of all samples except 1 (62 repeats in total so max is 62)
  tmp <-data.frame("gene_id"=c(unlist(gene_list))) %>% group_by(gene_id) %>% 
  dplyr::summarise(nr_occ=n())
  #ggplot(tmp, aes(x=nr_occ)) + geom_bar() + theme_point
  
  #Keep only genes that are significant in ALL sets to build model (+LOO crossval)
  nr_sets = length(colnames(cancer_z_table)[-1])
  consensus_genes_100[[cancertype]] <- tmp %>% filter(nr_occ==nr_sets) %>% pull(gene_id) #297 recurrent genes DLBCL, 931 PMBCL

}

#data.table::fwrite(left_join(data.frame(consensus_gene_id=consensus_genes_100), tail_genes, by=c("consensus_gene_id"="gene_id")),file = here::here("data_output","lymphoma_tail_consensusLOO.txt"),sep="\t",quote=F,na = "")

```


```{r}
### Now calculate nr of tail genes belonging to each type-specific biomarker tail set
tail_genes_FT_ndev <- data.frame()

for (cancer in c("DLBCL","PMBCL")){
tail_genes_FT_ndev_tmp <- tail_genes %>% filter(disease %in% c("CONTROL",cancer)) %>%
  filter(gene_id %in% (consensus_genes_100[[cancer]])) %>%
  group_by(sample) %>% dplyr::summarise(n_dev = n()) %>% 
  mutate(disease=gsub("_.*","",sample))
#make sure all control and cancer samples of comparison are still present
tail_genes_FT_ndev_tmp <- right_join(tail_genes_FT_ndev_tmp, 
                               data.frame(sample=unique(sample_annotation %>% filter(Cohort=="lymphoma") %>% filter(Abbreviation %in% c("CONTROL",cancer)) %>% pull(UniqueID))), by="sample") %>%
  mutate(disease=gsub("_.*","", sample)) %>% mutate(n_dev=ifelse(is.na(n_dev),0,n_dev)) %>%
  mutate(comparison=paste0(cancer,"vsCtrl")) %>% #add comparison to distinguish control outcomes
  mutate(disease=ifelse(disease=="CONTROL","CONTROL",disease))

tail_genes_FT_ndev <- rbind(tail_genes_FT_ndev, tail_genes_FT_ndev_tmp)
}

### add the across-cancer biomarker set
tail_genes_FT_ndev_tmp <- tail_genes %>% #filter(disease %in% c("CONTROL",cancer)) %>% #no need to filter samples
  filter(gene_id %in% (consensus_genes_100[["all"]])) %>%
  group_by(sample) %>% dplyr::summarise(n_dev = n()) %>% 
  mutate(disease=gsub("_.*","",sample))
#make sure all control and cancer samples of comparison are still present
tail_genes_FT_ndev_tmp <- right_join(tail_genes_FT_ndev_tmp, 
                               data.frame(sample=unique(sample_annotation %>% filter(Cohort=="lymphoma") %>% pull(UniqueID))), by="sample") %>%
  mutate(disease=gsub("_.*","", sample)) %>% mutate(n_dev=ifelse(is.na(n_dev),0,n_dev)) %>%
  mutate(comparison=paste0("allvsCtrl")) %>% #add comparison to distinguish control outcomes
  mutate(disease=ifelse(disease=="CONTROL","CONTROL","all"))

tail_genes_FT_ndev <- rbind(tail_genes_FT_ndev, tail_genes_FT_ndev_tmp)

data <- tail_genes_FT_ndev %>% mutate(cancer=ifelse(disease=="CONTROL", F, T)) %>% 
  mutate(disease=factor(disease)) %>% 
  mutate(disease=relevel(disease,ref="CONTROL"))

ggplot(data, aes(x=disease,y=n_dev, fill=disease)) +
  geom_boxplot(outlier.size = 0.6) + scale_y_log10() +
  geom_jitter(size=0.6, height=0,width=0.1,color="grey") +
  mytheme + labs(y="number of biomarker tail genes",x="") +
  theme(legend.position = "none", axis.text.x=element_text(angle=90, hjust=1,vjust=0.5)) +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488","all"="black")) +
  facet_wrap(~comparison, scales="free_x", nrow=1)

ggsave(filename="../figures/lymphoma_biomarkertail_boxplot_logscale.pdf", plot=last_plot(), width=5.8, height=6.1, units="cm")

ggplot(data %>% filter(comparison!="allvsCtrl"), aes(x=disease,y=n_dev, fill=disease)) +
  geom_boxplot(outlier.size = 0.6) + scale_y_log10() +
  geom_jitter(size=0.6, height=0,width=0.1,color="grey") +
  mytheme + labs(y="number of biomarker tail genes",x="") +
  theme(legend.position = "none") +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488","Ca"="black")) +
  facet_wrap(~comparison, scales="free_x", nrow=1)

ggsave(filename="../figures/lymphoma_biomarkertail_boxplot_logscale_withoutAll.pdf", plot=last_plot(), width=5.8, height=6.1, units="cm")

ggplot(data, aes(x=disease,y=n_dev, fill=disease)) +
  geom_boxplot(outlier.size = 0.6) +
  geom_jitter(size=0.6, height=0,width=0.1,color="grey") +
  mytheme + labs(y="number of biomarker tail genes",x="") +
  theme(legend.position = "none", axis.text.x=element_text(angle=90, hjust=1,vjust=0.5)) +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488","all"="black")) +
  facet_wrap(~comparison, scales="free_x",nrow=1)

ggsave(filename="../figures/lymphoma_biomarkertail_boxplot.pdf", plot=last_plot(), width=5.8, height=6.1, units="cm")

ggplot(data %>% filter(comparison!="allvsCtrl"), aes(x=disease,y=n_dev, fill=disease)) +
  geom_boxplot(outlier.size = 0.6) +
  geom_jitter(size=0.6, height=0,width=0.1,color="grey") +
  mytheme + labs(y="number of biomarker tail genes",x="") +
  theme(legend.position = "none") +
  scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488","all"="black")) +
  facet_wrap(~comparison, scales="free_x",nrow=1)

ggsave(filename="../figures/lymphoma_biomarkertail_boxplot_withoutAll.pdf", plot=last_plot(), width=5.8, height=6.1, units="cm")

my_comparisons=list(c("CONTROL","Cancer"), c("CONTROL","DLBCL"), c("CONTROL","PMBCL"))
#ggpubr::compare_means(n_dev ~ disease, data = data, method = "wilcox.test",p.adjust.method = "holm")
ggpubr::compare_means(n_dev ~ disease, data = data, method = "wilcox.test",p.adjust.method = "BH")

### add significance
wilcox.test(n_dev ~ disease, data=data %>% filter(comparison=="allvsCtrl"))
wilcox.test(n_dev ~ disease, data=data %>% filter(comparison=="DLBCLvsCtrl"))
wilcox.test(n_dev ~ disease, data=data %>% filter(comparison=="PMBCLvsCtrl"))

### rstatix significance
#rstatix::kruskal_test(data=data, n_dev ~ disease) #no KW test -> other control group values
#rstatix::kruskal_effsize(data=data, n_dev ~ disease)
#rstatix::kruskal_effsize(data=data, n_dev ~ disease,ci=T)
data_tmp <- data %>% mutate(disease=paste(disease)) #mutate disease column to string (otherwise error when performing wilcoxon tests because factors are missing)
rstatix::wilcox_test(data=data_tmp %>% filter(comparison=="allvsCtrl"), n_dev ~ disease)
#rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="allvsCtrl"), n_dev ~ disease)
rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="allvsCtrl"), n_dev ~ disease,ci=T)
rstatix::wilcox_test(data=data_tmp %>% filter(comparison=="DLBCLvsCtrl"), n_dev ~ disease)
#rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="DLBCLvsCtrl"), n_dev ~ disease)
rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="DLBCLvsCtrl"), n_dev ~ disease,ci=T)
rstatix::wilcox_test(data=data_tmp %>% filter(comparison=="PMBCLvsCtrl"), n_dev ~ disease)
#rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="PMBCLvsCtrl"), n_dev ~ disease)
rstatix::wilcox_effsize(data=data_tmp %>% filter(comparison=="PMBCLvsCtrl"), n_dev ~ disease,ci=T)

```

```{r}
require(pROC)
list_roc <- list()
for (cancertype in c("DLBCL","PMBCL")){
  data <- tail_genes_FT_ndev %>% mutate(cancer=ifelse(disease=="CONTROL", F, T)) %>% filter(comparison==paste0(cancertype,"vsCtrl"))
  k <- dim(data)[1]
  predictions <- c()
  for (i in 1:k) {
    model <- arm::bayesglm(cancer~n_dev, family="binomial", data=data[-i,])
    model <- glm(cancer~n_dev, family="binomial", data=data[-i,])
    predictions <- c(predictions, stats::predict(model, newdata=data[i,], type="response"))
  }
  list_roc[[cancertype]] <- roc(response=data$cancer,predictor= predictions) #add predictions for this comparison as a new list
  #data.frame(resp=data$cancer, pred=predictions, types=data$disease)
}

#### AUC calculation on across-cancer biomarker tail genes
tail_genes_FT_ndev_all <- tail_genes %>% filter(gene_id %in% (consensus_genes_100[["all"]])) %>%
  group_by(sample) %>% dplyr::summarise(n_dev = n()) %>% mutate(disease=gsub("_.*","",sample))
#make sure all samples are still present:
tail_genes_FT_ndev_all <- right_join(tail_genes_FT_ndev_all, 
                               data.frame(sample=unique(sample_annotation %>% filter(Cohort=="lymphoma") %>% pull(UniqueID))), by="sample") %>% mutate(disease=gsub("_.*","", sample)) %>% mutate(n_dev=ifelse(is.na(n_dev),0,n_dev))

data <- tail_genes_FT_ndev_all %>% mutate(cancer=ifelse(disease=="CONTROL", F, T))

k <- dim(data)[1]
predictions <- c()
for (i in 1:k) {
  model <- glm(cancer~n_dev, family="binomial", data=data[-i,])
  predictions <- c(predictions, stats::predict(model, newdata=data[i,], type="response"))
}
#library(pROC)
roc_modAll <- roc(response=data$cancer,predictor= predictions)


# ggroc(list(DLBCL=list_roc[["DLBCL"]],PMBCL=list_roc[["PMBCL"]],all=roc_modAll), legacy.axes = TRUE) +
#     mytheme + coord_fixed(ratio=1) +
#     theme(legend.position="right", legend.title = element_blank()) +
#     scale_color_manual(values=c("DLBCL"="#BB5566","PMBCL"="#004488","all"="black")) +
#     labs(x = 'False positive rate', y = 'True positive rate',
#          title = "ROC curve LOOCV (cancer-specific gene sets)\ncancer ~ n_dev",
#          subtitle= paste0("AUCs: DLBCL=",round(auc(list_roc[["DLBCL"]]),4), " [",round(ci(list_roc[["DLBCL"]], of="auc")[1],4),"-",round(ci(list_roc[["DLBCL"]], of="auc")[3],4),"];\n", 
#                           "PMBCL=",round(auc(list_roc[["PMBCL"]]),4), " [",round(ci(list_roc[["PMBCL"]], of="auc")[1],4),"-",round(ci(list_roc[["PMBCL"]], of="auc")[3],4),"];\n", 
#                           "all=",round(auc(roc_modAll),4), " [",round(ci(roc_modAll, of="auc")[1],4),"-",round(ci(roc_modAll, of="auc")[3],4),"];\n"))

ggroc(list(DLBCL=list_roc[["DLBCL"]],PMBCL=list_roc[["PMBCL"]],all=roc_modAll), legacy.axes = TRUE) +
    mytheme + coord_fixed(ratio=1) +
    theme(legend.position="bottom", legend.title = element_blank()) +
    scale_color_manual(values=c("DLBCL"="#BB5566","PMBCL"="#004488","all"="black")) +
    labs(x = 'False positive rate', y = 'True positive rate')

ggsave(filename = "../figures/lymphoma_biomarkertail_ROC_LOO.pdf", plot=last_plot(), width=5.8, height=6.1, units="cm")

auc(list_roc[["DLBCL"]])
ci(auc(list_roc[["DLBCL"]]))
auc(list_roc[["PMBCL"]])
ci(auc(list_roc[["PMBCL"]]))
auc(roc_modAll)
ci(auc(roc_modAll))

str(consensus_genes_100)

#plot precision recall instead
# plot(precision ~ recall,
#          coords(list_roc[["DLBCL"]], "all", ret = c("recall", "precision"), transpose = FALSE),
#          type="l", ylim = c(0, 1))
# plot(precision ~ recall,
#          coords(list_roc[["PMBCL"]], "all", ret = c("recall", "precision"), transpose = FALSE),
#          type="l", ylim = c(0, 1))

```

```{r, eval=F}
require(biomaRt)
ensembl <- useEnsembl(biomart="ensembl",dataset="hsapiens_gene_ensembl",version=91)
genes_ens_full <- getBM(attributes=c('ensembl_gene_id','external_gene_name','description'), mart =ensembl)

#biomarker_overview = data.frame(ensembl_gene_id=c(consensus_genes_100[["all"]], consensus_genes_100[["DLBCL"]], consensus_genes_100[["PMBCL"]]), biomarkerset=c(rep("AllvsCtrl",length(consensus_genes_100[["all"]])), rep("DLBCLvsCtrl",length(consensus_genes_100[["DLBCL"]])), rep("PMBCLvsCtrl",length(consensus_genes_100[["PMBCL"]])))) %>% left_join(genes_ens_full)

biomarker_overview = data.frame(ensembl_gene_id=c( consensus_genes_100[["DLBCL"]], consensus_genes_100[["PMBCL"]]), biomarkerset=c(rep("DLBCLvsCtrl",length(consensus_genes_100[["DLBCL"]])), rep("PMBCLvsCtrl",length(consensus_genes_100[["PMBCL"]])))) %>% left_join(genes_ens_full)
data.table::fwrite(biomarker_overview,file = "../data/lymphoma_biomarkertail_list.txt", sep="\t",quote=F,row.names=F)
```

Number of biomarker tail genes related to disease stage?
```{r}
#tail_genes_FT_ndev %>% filter(comparison=="DLBCLvsCtrl")
#tail_genes_FT_ndev %>% filter(comparison=="PMBCLvsCtrl")

tail_genes_FT_ndev <- tail_genes_FT_ndev %>% left_join(sample_annotation, by=c("sample"="UniqueID")) %>% mutate(Stage=ifelse(disease=="CONTROL","CONTROL",Stage))
ggplot(tail_genes_FT_ndev %>% filter(comparison!="allvsCtrl"), aes(x=Stage, y=n_dev, color=disease)) +
  geom_jitter(size=0.5, height=0, width=0.1) +
  facet_wrap(~comparison,scales="free_y") +
  mytheme +
  theme(legend.position="none") +
  labs(x=NULL, y="number of biomarker tail genes") +
  scale_color_manual(values=c("CONTROL"="#DDAA33","DLBCL"="#BB5566","PMBCL"="#004488"))

ggsave(filename = "../figures/lymphoma_biomarkertail_withcontrol.pdf", plot=last_plot(), width=5.8, height=5.1, units="cm")

glm(formula = n_dev~Abbreviation+Stage, data=tail_genes_FT_ndev %>% filter(comparison=="DLBCLvsCtrl")) %>% summary()
#glm(formula = n_dev~Stage, data=tail_genes_FT_ndev %>% filter(comparison=="DLBCLvsCtrl")) %>% summary()

glm(formula = n_dev~Abbreviation+Stage, data=tail_genes_FT_ndev %>% filter(comparison=="PMBCLvsCtrl")) %>% summary()
#glm(formula = n_dev~Stage, data=tail_genes_FT_ndev %>% filter(comparison=="DLBCLvsCtrl")) %>% summary()
```


Overlap of biomarker tail genes
```{r}
### overlap biomarker genes with differential genes
DAA_genes <- data.table::fread("../data/differentialabundance_lymphoma.txt", data.table = F) %>% mutate(DA=ifelse((padj<0.05) & (log2FoldChange>1), "higher", ifelse((padj<0.05) & (log2FoldChange<(-1)), "lower", "no")))  #differentially abudant genes (DE)
table(consensus_genes_100[["PMBCL"]] %in% (DAA_genes %>% filter(DA!="no") %>% filter(cancertype=="PMBC") %>% pull(gene_id)))
table(consensus_genes_100[["DLBCL"]] %in% (DAA_genes %>% filter(DA!="no") %>% filter(cancertype=="DLBC") %>% pull(gene_id)))
table(consensus_genes_100[["all"]] %in% (DAA_genes %>% filter(DA!="no") %>% pull(gene_id)))

### how many of the PMBCL and DLBCL consensus biomarker genes are differentially abundant?
table(unique(c(consensus_genes_100[["PMBCL"]],consensus_genes_100[["DLBCL"]])) %in% (DAA_genes %>% filter(DA!="no") %>% pull(gene_id)))

### overlap plot of biomarker tail genes
library(eulerr)
fit <- eulerr::euler(list("DLBCL"=consensus_genes_100[["DLBCL"]], "PMBCL"=consensus_genes_100[["PMBCL"]], "all"= consensus_genes_100[["all"]]),
                       shape="ellipse")
p1<- plot(fit, quantities = TRUE,shape="ellipse", alpha=0.7, edges="white", #make it a bit more light
          legend=F,fill=c(DLBCL="#BB5566",PMBCL="#004488",all="grey"), main="overlap biomarker tail gene sets")
print(fit) #no residual errors 
print(p1)
ggsave("../figures/lymphoma_biomarkertail_overlap.pdf", plot=p1, useDingbats=F, width=5.8, height=6.1, units="cm")

jaccard <- function(a, b) { #calculate jaccard index
    intersection = length(intersect(a, b))
    union = length(a) + length(b) - intersection
    return (intersection/union)
}


require(GeneOverlap)
#test overlap via GeneOverlap object
go.obj.DLPM <- newGeneOverlap(consensus_genes_100[["DLBCL"]],
                         consensus_genes_100[["PMBCL"]],
                         genome.size=length(unique(tail_genes$gene_id))) #all tail genes
go.obj.DLPM <- testGeneOverlap(go.obj.DLPM)
#go.obj.up #The GeneOverlap class formulates the problem as testing whether two variables are independent, which can be represented as a contingency table, and then uses Fisher’s exact test to find the statistical significance. Here the P-value is zero, which means the overlap is highly significant. The Fisher’s exact test also gives an odds ratio which represents the strength of association. If an odds ratio is equal to or less than 1, there is no association between the two lists. If the odds ratio is much larger than 1, then the association is strong. The class also calculates the Jaccard index which measures the similarity between two lists. The Jaccard index varies between 0 and 1, with 0 meaning there is no similarity between the two and 1 meaning the two are identical. To show some more details, use the print function
print(go.obj.DLPM)
jaccard(consensus_genes_100[["DLBCL"]], consensus_genes_100[["PMBCL"]])

go.obj.allPM <- newGeneOverlap(consensus_genes_100[["PMBCL"]],
                         consensus_genes_100[["all"]],
                         genome.size=length(unique(tail_genes$gene_id))) #all genes (19847)
go.obj.allPM <- testGeneOverlap(go.obj.allPM)
print(go.obj.allPM)
jaccard(consensus_genes_100[["all"]], consensus_genes_100[["PMBCL"]])

go.obj.allDL <- newGeneOverlap(consensus_genes_100[["DLBCL"]],
                         consensus_genes_100[["all"]],
                         genome.size=length(unique(tail_genes$gene_id))) #all tail genes?
go.obj.allDL <- testGeneOverlap(go.obj.allDL)
print(go.obj.allDL)
jaccard(consensus_genes_100[["DLBCL"]], consensus_genes_100[["all"]])

```


Internal validation: what if we randomly shuffle sample names (reassign samples to control or cancer group)? + determine consensus biomarker tail genes (leave-one-out Fisher's exact test) of (all) cancer vs control
```{r, warning=F}
#cohort PRAD, UCEC, OV, CONTROL
rm(tail_genes, z_scores_logscale_lymphoma,z_scores_logscale_lymphoma_normal_notinref,z_scores_LOO)

require(pROC)
require(ggpubr)

set.seed(124)
for (i in 1:20){
#for (i in 1:2){
  print(paste0("Iteration: ",i))
  lymphoma_normcounts <- data.table::fread("../data/lymphoma_normcounts.txt", data.table=F)
  
  ## reorder sample names (and thus cancer/normal type) randomly
  ordersamples_random <- sample(lymphoma_normcounts %>% dplyr::select(-gene_id) %>% colnames(.), replace=F)
  #print(grep("CONTROL",ordersamples_random[1:22],value=T))
  print(paste0(length(grep("CONTROL",ordersamples_random[1:22],value=T)), " original control samples in reference"))
  colnames(lymphoma_normcounts) <- c("gene_id",ordersamples_random)
  
  #calculate z scores for all genes in every sample based on new reference
  z_scores_LOO_random <- data.frame(gene_id = lymphoma_normcounts %>% pull(gene_id))
  for (samp in (colnames(lymphoma_normcounts)[-1])) {
    #print(samp)
    z_scores_LOO_random_tmp <- tail_z_score(lymphoma_normcounts %>% dplyr::select(c("gene_id",all_of(samp))) %>% column_to_rownames("gene_id") , 
                                                                    lymphoma_normcounts %>% dplyr::select(c("gene_id",grep("CONTROL",colnames(lymphoma_normcounts),value =T))) %>% 
                                                                      dplyr::select(-any_of(samp)) %>% #remove sample itself from ref
                                                                      column_to_rownames("gene_id"))
    
    z_scores_LOO_random <- left_join(z_scores_LOO_random, z_scores_LOO_random_tmp %>% rownames_to_column("gene_id"), by="gene_id")
  }

  rm(z_scores_LOO_random_tmp,samp)
  
  tail_genes <- left_join(z_scores_LOO_random %>% pivot_longer(names_to="sample", values_to="zscores",-"gene_id") %>% filter(!is.na(zscores)) %>% 
                          filter(abs(zscores)>3), 
                        lymphoma_normcounts %>% #add normalized counts (with same random reshuffled names)
                          pivot_longer(names_to="sample",values_to="ncounts",-"gene_id"),
                        by=c("gene_id"="gene_id","sample")) %>% filter(ncounts>=40) 
  
  #significant difference between new control group and cancer types?
  tmp2 <- tail_genes %>% group_by(sample) %>% dplyr::summarise(n_dev = n()) %>% mutate(disease=gsub("_.*","",sample))
  #make sure no sample is missing
  tmp2 <- tmp2 %>% full_join(sample_annotation %>% filter(Cohort=="lymphoma") %>% dplyr::select(c("sample"=UniqueID,"disease"=Abbreviation)), by=c("sample","disease")) %>% mutate(n_dev=ifelse(is.na(n_dev),0,n_dev))

  library(ggpubr)
  my_comparisons <- list( c("CONTROL", "DLBCL"),c("CONTROL", "PMBCL"))
  print(ggplot(tmp2, aes(x=disease,y=n_dev)) +
    geom_boxplot(outlier.size = 0.75,aes(fill=disease)) +
    geom_jitter(height=0, width=0.1,size=0.5,color="grey")+
    mytheme + theme(legend.position="none") +
    scale_fill_manual(values=c("CONTROL"="#DDAA33","DLBCL"="white","PMBCL"="white"))+
    scale_y_log10(labels=full_nr) +
    labs(y="tail genes",x="") +
    ggpubr::stat_compare_means(method="kruskal",label.y=4, size = 3) +  # Add global p-value
    ggpubr::stat_compare_means(comparisons=my_comparisons, method = "wilcox.test", size = 3)) # Pairwise comparison against reference
  #print(ggpubr::compare_means(n_dev ~ disease, data = tmp2,method = "wilcox.test",p.adjust.method = "BH"))
  print(kruskal.test(n_dev ~ disease, data = tmp2)) # significant difference between Disease types
  
}

##results of 20 iterations with filtering for Tail genes before Fisher test:
#c(23,1,24,5,3,9,8,0,1,7,14,28,1,0,4,0,30,2,27,0)
#number of controls in new control group
#c(6,5,10,7,7,9,9,8,4,10,6,9,7,9,6,8,4,11,7,8)

## get original normalized counts (with correct sample order)
lymphoma_normcounts <- data.table::fread("../data/lymphoma_normcounts.txt", data.table=F)

rm(tail_genes_random,z_scores_LOO_random, tail_genes, tail_genes_FT_ndev)

```